pulsar/producer_partition.go

```diff
@@ -898,16 +911,17 @@ func (p *partitionProducer) writeData(buffer internal.Buffer, sequenceID uint64,
 	default:
 		now := time.Now()
 		ctx, cancel := context.WithCancel(context.Background())
-		p.pendingQueue.Put(&pendingItem{
+		item := pendingItem{
 			ctx:          ctx,
 			cancel:       cancel,
 			createdAt:    now,
 			sentAt:       now,
 			buffer:       buffer,
 			sequenceID:   sequenceID,
 			sendRequests: callbacks,
-		})
-		p._getConn().WriteData(ctx, buffer)
+		}
+		p.pendingQueue.Put(&item)
+		p.writeChan <- &item
```

Comment:
The `writeChan` is a zero buffer channel. So if the client is reconnecting, the goroutine in line 564 will be stuck here. It couldn't timeout the message in the `dataChan`.